CodeMaid/Logic/Cleaning/FileHeaderLogic.cs

```diff
@@ -52,22 +58,142 @@ private FileHeaderLogic(CodeMaidPackage package)
+            var headerTemplate = GetFileHeaderFromSettings(textDocument);
+            if (string.IsNullOrWhiteSpace(headerTemplate))
             {
                 return;
             }
 
+            // Match against the file header based on lines
+            var headerLineLength = headerTemplate
+                .Split(new string[] { Environment.NewLine }, StringSplitOptions.None)
+                .Length;
+
+            var documentLineLength = textDocument.EndPoint.Line;
             var cursor = textDocument.StartPoint.CreateEditPoint();
-            var existingFileHeader = cursor.GetText(settingsFileHeader.Length);
+            var existingFileHeader = cursor.GetLines(1, Math.Min(headerLineLength, documentLineLength) + 1);
+
+            var variables = GetReplacementVariables(textDocument);
+            var headerPattern = BuildHeaderMatchPattern(variables, headerTemplate);
+            var matchResult = headerPattern.Match(existingFileHeader);
+            var updatedHeader = BuildHeader(variables, headerTemplate, matchResult);
 
-            if (!existingFileHeader.StartsWith(settingsFileHeader))
+            if (!existingFileHeader.StartsWith(updatedHeader))
             {
-                cursor.Insert(settingsFileHeader);
+                // delete the existing header that matches our pattern
+                if (matchResult.Success)
+                {
+                    var endOfHeader = textDocument.StartPoint.CreateEditPoint();
+                    endOfHeader.LineDown(headerLineLength);
+                    cursor.Delete(endOfHeader);
+                }
+
+                cursor.Insert(updatedHeader);
                 cursor.Insert(Environment.NewLine);
             }
         }
 
+        /// <summary>
+        /// Builds a regular expression that matches the file header based on the provided template
+        /// </summary>
+        /// <param name="variables">Variables that can have values substituted in the header</param>
+        /// <param name="headerTemplate">Template possibly containing variables</param>
+        /// <returns></returns>
+        private Regex BuildHeaderMatchPattern(FileHeaderVariable[] variables, string headerTemplate)
+        {
+            if (_cachedTemplates.TryGetValue(headerTemplate, out var cachedRegex))
+            {
+                return cachedRegex;
+            }
+
+            var resultBuilder = new StringBuilder();
+            resultBuilder.Append("^");
+            resultBuilder.Append(Regex.Escape(headerTemplate));
+
+            // Tolerate mixed line endings
+            resultBuilder.Replace(@"\r", "");
+            resultBuilder.Replace(@"\n", @"\r?\n");
+
+            foreach (var variable in variables)
+            {
+                // Because the settingsFileHeader was regex escaped above,
+                // we need to escape the variable name before replacing it with a regex value.
+                var escapedName = Regex.Escape("$" + variable.Name + "$");
```

Comment:
The reason I went with $VARIABLE$ was because other tools like ReSharper used it. Switching to % would still mean the input file header needs to be regex escaped, but the variable name would be the same escaped or unescaped. In my mind escaping both seems more consistent, but up to you on preference here

Reply:
I'm fine with $VARIABLE$ if that's a convention used elsewhere, that'll make things easier for people. 👍